src/components/componentContainer.ts

```diff
@@ -6,90 +6,77 @@ module Plottable {
+
+    constructor() {
+      super();
+      this._detachCallback = (component: Component) => this.remove(component);
+    }
 
     public anchor(selection: D3.Selection) {
       super.anchor(selection);
-      this.components().forEach((c) => c.anchor(this._content));
+      this._components().forEach((c) => c.anchor(this._content));
       return this;
     }
 
     public render() {
-      this._components.forEach((c) => c.render());
+      this._components().forEach((c) => c.render());
       return this;
     }
 
     /**
-     * Removes the specified Component from the ComponentContainer
-     * 
-     * @param c Component the Component to remove.
+     * Checks whether the specified Component is in the ComponentContainer.
      */
-    public remove(c: Component) {
-      var removeIndex = this._components.indexOf(c);
-      if (removeIndex >= 0) {
-        this.components().splice(removeIndex, 1);
-        this.redraw();
-      }
+    public has(component: Component) {
+      return this._components().indexOf(component) >= 0;
     }
 
-    /**
-     * Adds the specified Component to the ComponentContainer.
-     * 
-     * @param c Component the component to add
-     * @param prepend boolean whether the component should be prepended to the componentContainer or not.
-     */
-    public add(c: Component, prepend = false): boolean {
-      if (!c || this._components.indexOf(c) >= 0) {
-        return false;
-      }
-
-      if (prepend) {
-        this.components().unshift(c);
-      } else {
-        this.components().push(c);
-      }
-      c._parent(this);
+    protected _adoptAndAnchor(component: Component) {
+      component.parent(this);
+      component.onDetach(this._detachCallback);
       if (this._isAnchored) {
-        c.anchor(this._content);
+        component.anchor(this._content);
       }
-      this.redraw();
-      return true;
     }
 
     /**
-     * Returns a list of components in the ComponentContainer.
-     *
-     * @returns {Component[]} the contained Components
+     * Removes the specified Component from the ComponentContainer.
      */
-    public components(): Component[] {
-      return this._components;
+    public remove(component: Component) {
+      if (this.has(component)) {
+        component.offDetach(this._detachCallback);
+        this._remove(component);
+        component.detach();
+        this.redraw();
+      }
+      return this;
     }
 
     /**
-     * Returns true iff the ComponentContainer is empty.
-     *
-     * @returns {boolean} Whether the calling ComponentContainer is empty.
+     * Carry out the actual removal of a Component.
+     * Implementation dependent on the type of container.
+     * 
+     * @return {boolean} true if the Component was successfully removed, false otherwise.
      */
-    public empty() {
-      return this._components.length === 0;
+    protected _remove(component: Component) {
+      return false;
     }
```

Comment:
See above note. `_remove()` is an internal method that returns a boolean based on whether or not the removal was successful (similar to other list-like APIs).